src/fast/u192.rs

```diff
@@ -0,0 +1,57 @@
+use std::mem::transmute;
+use std::ops::{Add, Mul};
+
+// This was started by the output of the uint crate,
+// then heavily reduced to only the parts we need
+// (which was a significant optimization, especially
+// in the mul which carried out the full multiplication
+// and discarded the top-half)
+//
+/// Little-endian large integer type
+#[derive(Copy, Clone, PartialEq, Eq, Hash, Debug)]
+pub struct U192(pub [u64; 3]);
+
+impl Mul for U192 {
+    type Output = Self;
+    fn mul(self, other: Self) -> Self {
+        // The generated implementation of this method was 360 lines long!
+        let me = &self.0;
+        let you = &other.0;
+
+        let mult = |m: usize, y: usize| unsafe {
+            transmute::<_, (u64, u64)>(me[m] as u128 * you[y] as u128)
+        };
+
+        let (r0, r1) = mult(0, 0);
+        let (low, hi0) = mult(1, 0);
+        let (r1, overflow0) = low.overflowing_add(r1);
+        let (low, hi1) = mult(0, 1);
+        let (r1, overflow1) = low.overflowing_add(r1);
+
+        let r2 = (hi0 + overflow0 as u64)
+            .wrapping_add(hi1 + overflow1 as u64)
+            .wrapping_add(me[2].wrapping_mul(you[0]))
+            .wrapping_add(me[1].wrapping_mul(you[1]))
+            .wrapping_add(me[0].wrapping_mul(you[2]));
```

Comment:
I haven't really gone through what's happening here; I'll do that tomorrow.

Reply:
Reminder: double check this math.

I think it's likely to be correct because I ran with parallel implementations of the generated `U192` across all the tests, including the stress test / benchmark, and never found any discrepancies. It would panic on any intermediate value not matching. Even so, it would be good to get a pair of eyes to make sure there are no rare edge cases missed here.

Reply:
I just went through this, and AFAICT, this is correct.